SMCCC_ARCH_WORKAROUND_1

jenswi-linaro opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jforissier commented on core/arch/arm/kernel/thread_private.h:
```diff
@@ -228,6 +228,8 @@ void thread_set_irq_sp(vaddr_t sp);
 void thread_set_fiq_sp(vaddr_t sp);
 #endif /*ARM32*/
 
+unsigned long thread_smc(unsigned long func_id, unsigned long arg);
```
How about moving this elsewhere and making it more generic (i.e., accept more arguments), so that it may be used by platform code? We already have a use case as a replacement for https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/plat-ls/ls_hw_unq_key_a64.S.

jenswi-linaro replied:
Sure. I'd still want to keep it in this simple form, but accepting four arguments instead of only two.

The rational behind that being don't solve problems that doesn't exist. If we ever need a more complete function we can think about that then.

jforissier replied:
I agree. Four args is fine.

jforissier commented on core/arch/arm/kernel/thread.c:
I don't understand this. You're saying that it is enough to have an up-to-date ARM-TF to mitigate the CVE? How can this protect the EL0/EL1 transition? ARM-TF is not involved in this path, is it?

jenswi-linaro replied:
If the workaround isn't needed (fixed in newer cores or whatever) ARM-TF doesn't implement the workaround. Provided that ARM-TF is up to date.

jforissier replied:
Ah, OK. Now this makes sense.

etienne-lms commented on core/arch/arm/include/smccc.h:
```diff
@@ -0,0 +1,36 @@
+
+/*
+ * Describes features of SMC Calling Convention from v1.1
+ * See also https://developer.arm.com/-/media/developer/pdf/ARM%20DEN%200070A%20Firmware%20interfaces%20for%20mitigating%20CVE-2017-5715_V1.0.pdf
```
ref is deprecated.

Looks like latest seems https://developer.arm.com/-/media/developer/pdf/ARM_DEN_0070A_Firmware_interfaces_for_mitigating_CVE-2017-5715.pdf?revision=d20d1476-b770-4739-aebe-2d3147788e3f&la=en.

But I think we should rather point to 
https://developer.arm.com/-/media/developer/pdf/ARM_DEN_0070A_Firmware_interfaces_for_mitigating_CVE-2017-5715.pdf

etienne-lms commented on core/arch/arm/kernel/thread_a64.S:
```diff
@@ -655,7 +673,16 @@ eret_to_el0:
 	mrs	x0, tpidr_el1
+	sub	sp, sp, x0
+#endif
+
```
minor: empty line

jenswi-linaro replied:
That is intentional

etienne-lms commented on core/arch/arm/mm/tee_mmu.c:
```diff
@@ -360,18 +360,25 @@ static void asid_free(unsigned int asid)
+	if (sz)
+		return vm_map(utc, &va, sz, TEE_MATTR_PRW | TEE_MATTR_PERMANENT,
+			      mobj, offs);
 	return TEE_SUCCESS;
```
minor: maybe add an empty line

etienne-lms commented on core/arch/arm/kernel/thread_a64.S:
```diff
@@ -373,6 +373,17 @@ END_FUNC thread_unwind_user_mode
 		msr	vbar_el1, x1
 		isb
 
+
```
minor: remove empty line

etienne-lms commented on core/arch/arm/kernel/thread.c:
```diff
@@ -905,6 +913,15 @@ static void init_user_kcode(void)
 
 	core_mmu_get_user_va_range(&v, NULL);
 	thread_user_kcode_offset = thread_user_kcode_va - v;
+#if defined(CFG_CORE_WORKAROUND_SPECTRE_BP_SEC) && defined(ARM64)
```
minor: add an empty line

etienne-lms commented on core/arch/arm/kernel/thread.c:
```diff
@@ -152,7 +152,15 @@ thread_pm_handler_t thread_system_reset_handler_ptr;
+long thread_user_kdata_sp_offset;
+static uint8_t thread_user_kdata_page[
+	ROUNDUP(sizeof(thread_core_local), SMALL_PAGE_SIZE)]
+	__aligned(SMALL_PAGE_SIZE) __section(".nozi.kdata_page");
```
All active threads uses the same `thread_user_kdata_page`?

jenswi-linaro replied:
It's actually the cores that uses the same `thread_user_kdata_page`.

jenswi-linaro commented:
Update

jforissier commented on core/arch/arm/kernel/thread.c:
```diff
@@ -975,10 +1013,10 @@ static vaddr_t get_excp_vect(void)
 	case CORTEX_A72_PART_NUM:
 	case CORTEX_A73_PART_NUM:
 	case CORTEX_A75_PART_NUM:
```
Since we rely on ARM-TF for implementing the workaround, should we still keep this list of affected cores?

jenswi-linaro replied:
Yes, I see no harm in that.

jenswi-linaro commented:
Squashed and rebased, tags applied.
Note that the commits "core: provide special rw kernel page at EL0 " and "core: calculate size of special rx map at EL0" doesn't have any new tags.

jenswi-linaro commented:
So this didn't make it for the RC1.

jforissier commented:
@jenswi-linaro if we can't get review or ack tags for the kernel page mapping, then so be it! I'm willing to merge this anyway.

etienne-lms commented:
sorry sorry.
i though you could wait 3.1 tag for merging this. I would say looks-good-to-me, but as you asked me to have a look I wanted to take some time on it.
Targeting 3.1? If so tell me, i will spend some time on this today.

jforissier commented:
> Targeting 3.1?

Yes, definitely.

jenswi-linaro commented:
I've rebased it on master and squashed the small fix for "core: arm: add thread_smc()" (smc aarch32).

etienne-lms commented on core/arch/arm/kernel/thread.c:
minor: Maybe replace `Workaround` with `SMCCC workaround`, here and below.

jforissier commented:
The [checkpatch warnings](https://travis-ci.org/OP-TEE/optee_os/builds/365492781#L2197-L2212) can both be fixed:
- split the line > 80 characters
- use a Link: tag in the commit description -- such a tag won't trigger the "Possible unwrapped commit description" message

jenswi-linaro commented:
You mean splitting the debug print lines, don't we agree on keeping them intact for easier grep:ing?

lorc commented:
> You mean splitting the debug print lines, don't we agree on keeping them intact for easier grep:ing?

Actually linux coding style clearly states that long text strings should be leaved as is:


> Statements longer than 80 columns will be broken into sensible chunks, unless
> exceeding 80 columns significantly increases readability and does not hide
> information. Descendants are always substantially shorter than the parent and
> are placed substantially to the right. The same applies to function headers
> with a long argument list. However, never break user-visible strings such as
> printk messages, because that breaks the ability to grep for them.

jforissier commented:
> Actually linux coding style clearly states that long text strings should be leaved as is

Right, makes sense. It's unfortunate that checkpatch still reports those lines...

jenswi-linaro commented:
@etienne-lms, regarding "core: arm64: provide special rw kernel page at EL0". As you say it has nothing to do with arm32, why talk about a solution for that which already is in place?

Tag applied and commit messages updated.
Squashed and rebased the updates. Diff against 3aadb0c983a2bc675b1d0b307718f10561b9af34 to see the small changes.

etienne-lms commented:
`arm64:` ref in the 1st line of the commit comment is fine :), thanks.
